### PR TITLE
Add UIAA error types

### LINKED ISSUE #136: Add new error type for User-Interactive Authentication API responses
The UserInteractive endpoints are:
* `POST /_matrix/client/r0/register`
* `POST /_matrix/client/r0/account/password`
* `POST /_matrix/client/r0/account/deactivate`
* `POST /_matrix/client/r0/account/3pid/add`
* `DELETE /_matrix/client/r0/devices/{deviceId}`
* `POST /_matrix/client/r0/delete_devices`

I think it would be fair to define the error type for those six as an enum of either crate::error or r0::account::UIAAResponse

### PR DESCRIPTION
**Pull request by iinuwa**
Fixes #136.

Not ready to merge yet; should write some tests to verify it works.

### CODE REVIEW

**Comment by iinuwa** [comment]
Before this is finished, I'd at least like to write a couple tests for `UserInteractiveAuthenticationResponse::try_from_response()`. Do you have a preference on how granular these tests should be?

It may also be good to move all the UIAA stuff into a separate module or a sub-module of `r0::account`, since it's used outside of `r0::account` and it kind of overshadows the other two types in that module. What do you think?

**Comment by jplatte** [comment]
> Do you have a preference on how granular these tests should be?

No.

> It may also be good to move all the UIAA stuff into a separate module or a sub-module of `r0::account`, since it's used outside of `r0::account` and it kind of overshadows the other two types in that module. What do you think?

Sounds good!

**Comment by iinuwa** [comment]
Re-posting here to remind me that I need to do this later:

Now that I'm reading more about UIAA, it seems that all the fields in `AuthenticationData` need to be optional.
> Once a client receives the notificaton that the authentication stage has been completed, it should resubmit the request with an auth dict with just the session ID:
> ```json
> {
>   "session": "<session ID>"
> }
> ```

Cf. https://matrix.org/docs/spec/client_server/r0.6.0#fallback

**Comment by jplatte** [comment]
Can the fields be absent individually, or is it an all-or-nothing situation? In the latter case, a separate struct that is included as `#[serde(flatten)] field: Option<Fields>` would probably be the best solution.

**Comment by iinuwa** [comment]
There are three valid ways a client can send a UIAA request:
- session only (for response after fallback auth stage): `{"session": "xxxxxx"}`
- session with login type: `{"type": "m.login.dummy", "session": "xxxxxx"}`
- session with type and arbitrary parameters: `{"type": "m.login.password", "session": "xxxxxx", "password": "abc123", "identifier": {...} }`

Riot also sends an empty dict, or auth object with no session, type or parameters, but that's a bug since r0.5.0.

The only change that this would require is to change the `kind` field to be an `Option<String>`, resulting in:
```rust
pub struct AuthenticationData {
    /// The login type that the client is attempting to complete.
    #[serde(rename = "type")]
    pub kind: Option<String>,
    /// The value of the session key given by the homeserver.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub session: Option<String>,
    /// Parameters submitted for a particular authentication stage.
    #[serde(flatten)]
    pub auth_parameters: BTreeMap<String, serde_json::Value>,
}
```

Coincidentally, I think this would solve the problem of Riot not being able to connect to Timo's server because of this (until that bug is patched), since Ruma would still be able to deserialize `{}` as `AuthenticationData { kind: None, session: None, auth_parameters: BTreeMap<String, serde_json::Value>:::new() }` without throwing an error.

**Comment by jplatte** [comment]
Sounds good! Though are there potentially advantages to making the completely empty case irrepresentable (by having `AuthenticationData` be an enum)?

**Comment by iinuwa** [comment]
Yeah, technically the empty object is "against the spec" (though it seems the request schema described in section 5.3 and the Authentication Data schema conflict with each other, so who's to say who's right).

We could do something like:
```rust
pub enum AuthenticationData {
  DirectResponse {
    #[serde(rename = "type")]  
    kind: String,
    #[serde(skip_serializing_if = Option::is_none)]
    session: Option<String>,
    #[serde(flatten)]
    parameters: BTreeMap::<String, serde_json::Value>,
  },
  FallbackResponse { session: String  }
}
```

**Comment by jplatte** [comment]
If the spec is in conflict with itself basically, I would be fine with merging either version. But considering the empty object being sent by Riot is getting removed, I guess this version is slightly better.

File: src/r0/uiaa.rs [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+                ));
+            }
+        }
+
+        match crate::error::Error::try_from_response(response) {
+            Ok(matrix_error) => Ok(UserInteractiveAuthenticationResponse::MatrixError(
+                matrix_error,
+            )),
+            Err(de_error) => Err(de_error),
+        }
```
Comment (jplatte) [inline src/r0/uiaa.rs:95]: This `match` could instead be `.map(UserInteractiveAuthenticationResponse::MatrixError)`, right?

Alternatively, `UserInteractiveAuthenticationResponse` (by the way, did we want to shorten the name?) could implement `From<Error>`, so the map becomes just `.map(From::from)`.

**Comment by iinuwa** [comment]
Yeah, the spec should be revised to make `session` required, since that's how it is used in practice. I think the enum we have works fine until then.

File: src/r0/uiaa.rs [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::collections::BTreeMap;
+
+    use ruma_api::EndpointError;
+    use serde_json::{from_value as from_json_value, json, to_value as to_json_value, Value};
```
Comment (jplatte) [inline src/r0/uiaa.rs:121]:
```suggestion
    use serde_json::{from_value as from_json_value, json, to_value as to_json_value, Value as JsonValue};
```

**Comment by jplatte** [comment]
:tada:

**Pull request merged**